Add `#env-var` macro for compile time definitions

callendorph opened the pull request:
This commit adds a new macro definition for accessing environment variables defined during the compilation process.

For example - if you wanted to pass the VERSION info into the build without hard coding it in the code, you could:

```
VERSION="1.2.3" stanza build
```

In your stanza code:

```
val version = #env-var(VERSION)
```

I've attempted to keep the syntax similar to the `#define`, `#if-defined` etc macros by using ids as the argument. There is probably an argument for allowing this to take an expression but I was worried about the side-effects of evaluating an expression during compilation. Let me know what you think.

I've tested this in OS-X for both the happy path and the two syntax error paths (`#env-var` and `#env-var()`) which successfully throw syntax errors.

jackbackrack commented:
i'm wondering if there's a more robust way to do this.  for example, perhaps this can be done by copying the map but adding a false and then using dead code elimination to prune out all the unnecssary object creation.  this hasn't shown much in the way of speed ups so far so it's inclusion should perhaps be delayed.

CuppoJava commented:
It can be done automatically by specializing functions based upon the context in which their return value is used. In this case, the compile could derive a version of `map` that is used whenever it can deduce that its return value is not used. Because the return value is not used, then it does not need to create any output objects. Once that is recursively applied to everything in `map`, it will be identical to `do`.

callendorph committed:
Add `#env-var` macro for compile time definitions

This commit adds a new macro definition for accessing environment variables
defined during the compilation process.

For example - if you wanted to pass the VERSION info into the build
without hard coding it in the code, you could:

```
VERSION="1.2.3" stanza build
```

In your stanza code:

```
val version = #env-var(VERSION)
```

I've attempted to keep the syntax similar to the `#define`, `#if-defined` etc macros by using ids as the argument.

CuppoJava commented:
Thanks! Looks great!
